odoo/models.py

```diff
@@ -4843,7 +4841,18 @@ def sudo(self, user=SUPERUSER_ID):
             The returned recordset has the same prefetch object as ``self``.
 
         """
-        return self.with_env(self.env(user=user))
+        if not isinstance(flag, bool):
+            _logger.warning("deprecated use of sudo(user), use with_user(user) instead", stack_info=True)
```

Comment:
For the record, we don't keep backwards-compatibility with explicit kwarg `user`, as in:
```python
   self.sudo(user=uid)
```
because this usage is quite rare and would require increasing the technical debt (extra/useless param + complexity).

Those cases will have to be explicitly converted to `with_user(user=uid)` - the error will be immediate and obvious enough.